File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "null,";

        StringBuilder sb = new StringBuilder();
        sb.append(root.val);
        sb.append(",");
        sb.append(serialize(root.left));
        sb.append(serialize(root.right));

        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Queue<String> queue = new LinkedList<>(Arrays.asList(data.split(",")));
        return dfs(queue);
    }

    private TreeNode dfs(Queue<String> queue) {
        String cur = queue.poll();

        if (cur.equals("null")) return null;

        TreeNode node = new TreeNode(Integer.parseInt(cur));
        node.left = dfs(queue);
        node.right = dfs(queue);

        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
	// Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        if (root == null) return "null,"

        return buildString {
            append(root.`val`)
            append(",")
            append(serialize(root.left))
            append(serialize(root.right))
        }
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        val queue = ArrayDeque(data.split(","))
        return dfs(queue)
    }

    private fun dfs(queue: ArrayDeque<String>): TreeNode? {
        val cur = queue.removeFirst()

        if (cur == "null") return null

        val node = TreeNode(cur.toInt())
        node.left = dfs(queue)
        node.right = dfs(queue)

        return node
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * var ser = Codec()
 * var deser = Codec()
 * var data = ser.serialize(longUrl)
 * var ans = deser.deserialize(data)
 */